lib/pbench/test/unit/agent/test_tool_meister.py

```diff
@@ -0,0 +1,1165 @@
+            tool.stop()
+
+    @staticmethod
+    @pytest.mark.parametrize("delay_cnt", [99, 100, 101])
+    def test_stop_delayed_pid_file(delay_cnt, monkeypatch, caplog):
+        """Verify the behavior of waiting for the pid file to show up.
+
+        The algorithm used by `.stop()` is to wait 10 seconds for the pid file
+        to come into existence by sleeping 1/10 of a second, and checking for
+        the file to exist, each time through the loop for 100 iterations of
+        the loop.
+
+        We test the algorithm by executing the mock making the file come into
+        existence on the 99th call to Path.exists() (resulting in no warning
+        message), on the 100th call to Path.exists() (again no warning
+        message), and then on the 101st call to Path.exists() (which won't
+        ever happen, so a warning message will be issued).
+        """
+        tool = TestTransientTool.setup_transient_tool(
+            f"test_stop_delayed_pid_file:delay_cnt={delay_cnt}"
+        )
+
+        exists_cnt = {}
+
+        class MockedPathDelayedExists(MockedPath):
+            """Minimal mocked Path object extended to delay reporting a given file
+            exists."""
+
+            def __init__(
+                self, exists: bool = True, path: str = "/foo/bar", delay_cnt: int = 0
+            ):
+                super().__init__(exists=exists, path=path)
+                self._delay_cnt = delay_cnt
+                self._exists_cnt = 0
+
+            def exists(self) -> bool:
+                self._exists_cnt += 1
+                nonlocal exists_cnt
+                exists_cnt[self._path] = self._exists_cnt
+                if self._exists_cnt >= self._delay_cnt:
+                    return self._exists
+                else:
+                    return False
+
+            def __truediv__(self, path: str):
+                return MockedPathDelayedExists(
+                    self._exists, f"{self._path}/{path}", delay_cnt=self._delay_cnt
+                )
+
+        the_tool_dir = MockedPathDelayedExists(delay_cnt=delay_cnt)
+        the_pid_file = f"{the_tool_dir}/test-tool/test-tool.pid"
+        tool.tool_dir = the_tool_dir
+
+        def noop(*args, **kwargs):
+            return
+
+        monkeypatch.setattr("time.sleep", noop)
+
+        tool.start_process = "running"
+
+        tool.stop()
+
+        assert (
+            len(exists_cnt) == 1
+        ), f"Problem - more than one file used, {exists_cnt.keys()!r}"
+        assert exists_cnt[the_pid_file] == min(delay_cnt, 100)
+        expected_logs = 1 if delay_cnt <= 100 else 2
+        assert len(caplog.record_tuples) == expected_logs
+        # Note the info message is always the "last" log message emitted by
+        # the `.stop()` method.
+        assert caplog.record_tuples[-1][1] == logging.INFO
+        assert (
+            caplog.record_tuples[-1][2]
+            == f"test-tool: stop_tool -- {tool.tool_script} --stop --dir={the_tool_dir} {tool.tool_opts}"
```

Comment:
It would be better if we didn't depend on `tool` for these values....

Reply:
The problem is, if we pull the values out of `tool` which `tool` is going to use to generate the string, then the assertion doesn't really test very much.  (The phrase "GIGO" leaps to mind.)

Reply:
Let's continue any further refactoring and test enhancements in PR #2725.